### PR TITLE
consensus: remove logic to unlock block on 2/3 prevote for nil

### PR DESCRIPTION
**Pull request by williambanfield**
## What does this change do?

This change removes the logic for a validator to 'unlock' its locked block if it sees > 2/3 prevotes for nil in a round. This condition is not in the algorithm description in the Tendermint paper.

This logic was present in both `enterPrecommit` and `addVote`. The logic in `addVote` has been updated to only unlock on receipt of 2/3 prevotes for non-nil values and the logic in `enterPrecommit` has been completely removed. 

This change also updates the tests to add a new test for the unlock condition. As a cosmetic fixup, the test file was also updated to match the comments with the current names of the `Lock` suite of tests. Finally, the `Relock` test was updated to use the lock channels to ensure the unlock/lock logic remained correct despite this change. The test for the old behavior was removed. 

## Notes

* The logic for a validator to prevote its own locked block regardless of proposal is maintained during this change. It will be updated in a future pull request.

closes: #6849

### CODE REVIEW

**Comment by williambanfield** [comment]
cc @zeu5

File: internal/consensus/state.go [hunk 1]
```diff
@@ -1408,21 +1407,9 @@ func (cs *State) enterPrecommit(height int64, round int32) {
 		panic(fmt.Sprintf("this POLRound should be %v but got %v", round, polRound))
 	}
 
-	// +2/3 prevoted nil. Unlock and precommit nil.
+	// +2/3 prevoted nil.
```
Comment (williambanfield) [inline internal/consensus/state.go:1410]:
```suggestion
	// +2/3 prevoted nil. Precommit nil.
```

File: internal/consensus/state.go [hunk 2]
```diff
@@ -2070,20 +2057,16 @@ func (cs *State) addVote(vote *types.Vote, peerID types.NodeID) (added bool, err
...
 				(cs.LockedRound < vote.Round) &&
 				(vote.Round <= cs.Round) &&
-				!cs.LockedBlock.HashesTo(blockID.Hash) {
+				!cs.LockedBlock.HashesTo(blockID.Hash) && len(blockID.Hash) != 0 {
 
-				cs.Logger.Debug("unlocking because of POL", "locked_round", cs.LockedRound, "pol_round", vote.Round)
+				cs.Logger.Debug("unlocking because of POL for non-nil block", "locked_round", cs.LockedRound, "pol_round", vote.Round)
 
 				cs.LockedRound = -1
 				cs.LockedBlock = nil
```
Comment (cmwaters) [inline internal/consensus/state.go:2072]: So we have 2/3+ prevotes for a block that's not the one we locked on and so we unlock. Do we lock on this new block in the same round and precommit it? Or precommit nil. It's not clear what action we are doing at this stage. On line 1432 it says to stage and precommit it but I don't see where we do that

**Review by cmwaters — APPROVED** [review]
Looks mostly good. I have a few questions / comments.

File: internal/consensus/state.go [hunk 3]
```diff
@@ -2070,20 +2057,16 @@ func (cs *State) addVote(vote *types.Vote, peerID types.NodeID) (added bool, err
 		prevotes := cs.Votes.Prevotes(vote.Round)
 		cs.Logger.Debug("added vote to prevote", "vote", vote, "prevotes", prevotes.StringShort())
 
-		// If +2/3 prevotes for a block or nil for *any* round:
+		// If +2/3 prevotes for a block for *any* round:
 		if blockID, ok := prevotes.TwoThirdsMajority(); ok {
 			// There was a polka!
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply internal/consensus/state.go]: Updated. Clear pose descriptions are likely more meaningful for these.

File: internal/consensus/state.go [hunk 4]
```diff
@@ -2070,20 +2057,16 @@ func (cs *State) addVote(vote *types.Vote, peerID types.NodeID) (added bool, err
...
-			// If we're locked but this is a recent polka, unlock.
 			// If it matches our ProposalBlock, update the ValidBlock
-
-			// Unlock if `cs.LockedRound < vote.Round <= cs.Round`
-			// NOTE: If vote.Round > cs.Round, we'll deal with it when we get to vote.Round
 			if (cs.LockedBlock != nil) &&
 				(cs.LockedRound < vote.Round) &&
 				(vote.Round <= cs.Round) &&
-				!cs.LockedBlock.HashesTo(blockID.Hash) {
+				!cs.LockedBlock.HashesTo(blockID.Hash) && len(blockID.Hash) != 0 {
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply internal/consensus/state.go]: done.

File: internal/consensus/state.go (hunk 2 above)
Comment (williambanfield) [reply internal/consensus/state.go:2078]: From my investigation, the unlock here is entirely useless. I'm planning on cleaning it up in #6850

If you notice the update to the TestStateLockPOLRelock, you can see we check for an unlock followed by a lock within the same consensus round. 
https://github.com/tendermint/tendermint/blob/c69f933d61962ffae5420b1cb476bfa2e5c96a1f/internal/consensus/state_test.go#L704

This means that the unlock step here is not doing anything since we're going to relock within the same round. It's therefore totally safe to remove.

File: internal/consensus/state_test.go [hunk 5]
```diff
@@ -745,55 +764,88 @@ func TestStateLockPOLUnlock(t *testing.T) {
...
+	signAddVotes(config, cs1, tmproto.PrecommitType, nil, types.PartSetHeader{}, vs2, vs3, vs4)
+	ensureNewTimeout(timeoutWaitCh, height, round, cs1.config.Precommit(round).Nanoseconds())
+
 	/*
-		Round2 (vs2, C) // B nil nil nil // nil nil nil _
-		cs1 unlocks!
+		Round 2:
+		The validator cs1 saw >2/3 precommits for nil in the previous round.
+		Send the validator >2/3 prevotes for nil and ensure that it did not
+		unlock its block at the end of the previous round.
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply internal/consensus/state_test.go:825]: It's just verifying that 2/3 precommits for nil don't impact the locked block. A check for that wasn't part of any test before and I think adding the coverage will help make sure that any changes to this logic don't accidentally introduce new incorrect behavior.

File: internal/consensus/state_test.go [hunk 6]
```diff
@@ -1196,41 +1247,35 @@ func TestProposeValidBlock(t *testing.T) {
...
 
 	signAddVotes(config, cs1, tmproto.PrecommitType, nil, types.PartSetHeader{}, vs2, vs3, vs4)
 
-	round += 2 // moving to the next round
+	round += 2 // increment by multiple rounds
 
 	ensureNewRound(newRoundCh, height, round)
 	t.Log("### ONTO ROUND 3")
 
-	ensureNewTimeout(timeoutWaitCh, height, round, cs1.config.Precommit(round).Nanoseconds())
```
[the remark this answers is not in the document]
Comment (williambanfield) [reply internal/consensus/state_test.go]: fixed.

File: internal/consensus/state.go [hunk 7]
```diff
@@ -1408,21 +1407,9 @@ func (cs *State) enterPrecommit(height int64, round int32) {
 		panic(fmt.Sprintf("this POLRound should be %v but got %v", round, polRound))
 	}
 
-	// +2/3 prevoted nil. Unlock and precommit nil.
+	// +2/3 prevoted nil. Precommit nil.
```
Comment (milosevic) [inline internal/consensus/state.go:1410]: Side comment: we should probably have a helper function that wraps `len(blockID.Hash) == 0`, for example `isVoteNil` or something like that, as code would read better.

Comment (milosevic) [reply internal/consensus/state.go:1410]: Also I am wondering what exactly the check at line 1405 brings? I don't see from correctness perspective entering precommit in case there is a polka for higher round. As this also depends on the way gossip is implemented and skipping round logic, we could end up in scenario where this condition is not true and then we would panic. As a more general rule of thumb, I think we should have clear description or even better invariant violation whenever we panic in the code.

File: internal/consensus/state.go [hunk 8]
```diff
@@ -1442,7 +1429,9 @@ func (cs *State) enterPrecommit(height int64, round int32) {
 		return
 	}
 
-	// If +2/3 prevoted for proposal block, stage and precommit it
+	// If greater than 2/3 of the voting power on the network prevoted for
+	// the proposed block, update our locked block to this block and issue a
+	// precommit vote for it.
```
Comment (milosevic) [inline internal/consensus/state.go:1434]: What about unlocking logic at lines 1460-1462? In case there is a non-nil polka for a block we don't have, we currently unlock. I think we should also remove these lines together with L1469. In this scenario it is sufficient to precommit nil and update `cs.ProposalBlock` and `cs.ProposalBlockParts` so we start fetching the block for which there is a polka.

File: internal/consensus/state.go [hunk 9]
```diff
@@ -2070,20 +2059,20 @@ func (cs *State) addVote(vote *types.Vote, peerID types.NodeID) (added bool, err
...
+		// Check to see if >2/3 of the voting power on the network voted for any non-nil block.
+		if blockID, ok := prevotes.TwoThirdsMajority(); ok && len(blockID.Hash) != 0 {
+			// Greater than 2/3 of the voting power on the network voted for some
+			// non-nil block
 
-			// Unlock if `cs.LockedRound < vote.Round <= cs.Round`
-			// NOTE: If vote.Round > cs.Round, we'll deal with it when we get to vote.Round
+			// If we locked a different block in an earlier round, unlock it.
+			// We are going to relock a new block when we precommit, so this unlock
+			// is only temporary.
```
Comment (milosevic) [inline internal/consensus/state.go:2069]: In my view, it would be cleaner completely getting rid of unlocking.

File: internal/consensus/state.go [hunk 10]
```diff
@@ -2070,20 +2059,20 @@ func (cs *State) addVote(vote *types.Vote, peerID types.NodeID) (added bool, err
 		prevotes := cs.Votes.Prevotes(vote.Round)
 		cs.Logger.Debug("added vote to prevote", "vote", vote, "prevotes", prevotes.StringShort())
 
-		// If +2/3 prevotes for a block or nil for *any* round:
-		if blockID, ok := prevotes.TwoThirdsMajority(); ok {
-			// There was a polka!
-			// If we're locked but this is a recent polka, unlock.
-			// If it matches our ProposalBlock, update the ValidBlock
+		// Check to see if >2/3 of the voting power on the network voted for any non-nil block.
```
Comment (milosevic) [inline internal/consensus/state.go:2062]: Side comment: we should probably add invariant check that `if len(blockID.Hash) != 0 ` at the beginning of this function, for example at line 1535. It is currently done at the caller, but we should probably have it explicitly also inside this function so we don't rely only on caller to do it properly.

File: internal/consensus/state.go (hunk 2 above)
Comment (milosevic) [reply internal/consensus/state.go:2078]: Agree this is not needed. Maybe it have been better removing unlock logic in this PR in general, not just in case where there is 2/3+ nil prevotes.

File: internal/consensus/state_test.go [hunk 11]
```diff
@@ -622,6 +622,8 @@ func TestStateLockPOLRelock(t *testing.T) {
 	require.NoError(t, err)
 	addr := pv1.Address()
 	voteCh := subscribeToVoter(cs1, addr)
+	unlockCh := subscribe(cs1.eventBus, types.EventQueryUnlock)
```
Comment (milosevic) [inline internal/consensus/state_test.go:625]: Would be good if we can git rid of Unlock event all together, but probably you want to do it in different PR.

File: internal/consensus/state_test.go [hunk 12]
```diff
@@ -705,31 +716,39 @@ func TestStateLockPOLRelock(t *testing.T) {
 	ensureNewRound(newRoundCh, height+1, 0)
 }
 
-// 4 vals, one precommits, other 3 polka at next round, so we unlock and precomit the polka
-func TestStateLockPOLUnlock(t *testing.T) {
+// TestStatePOLDoesNotUnlock tests that a validator maintains its locked block
+// despite receiving +2/3 nil prevotes and nil precommits from other validators.
```
Comment (milosevic) [inline internal/consensus/state_test.go:720]: Receiving of +2/3 nil precommits was never the condition to unlock. So maybe it would be good removing it from the comment.

File: internal/consensus/state_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (milosevic) [reply internal/consensus/state_test.go:825]: See comment above: nil precommits are not related to the unlocking rule.

**Review by milosevic — APPROVED** [review]
Looks good to me. I would have slight preference for removing unlock rules all together in one PR, and not just for the +2/3 prevote nil case, but it's also fine if it is done as part of follow up PR.

File: internal/consensus/state.go (hunk 7 above)
Comment (williambanfield) [reply internal/consensus/state.go:1410]: There's a helper for nil block checking. I just renamed it now to `blockID.IsNil()` and updated the code in `state.go` to use it.

File: internal/consensus/state.go (hunk 8 above)
Comment (williambanfield) [reply internal/consensus/state.go:1434]: Fair enough, I was planning to remove this in a follow up PR, but I'm happy to just remove it all now.

File: internal/consensus/state.go (hunk 10 above)
Comment (williambanfield) [reply internal/consensus/state.go:2053]: I think I'm a little confused by which function you're referring to. This code here is within `addVote`, which doesn't seem to require that there be a non-nil polka yet.

Comment (milosevic) [reply internal/consensus/state.go:2053]: It is `enterCommit` function, line 1535.

Comment (williambanfield) [reply internal/consensus/state.go:2053]: I see, I'm going to create a github issue to track this. I'd prefer to keep this commit somewhat focused so that it's easier to roll back in the event that it causes any issue.

File: internal/consensus/state_test.go (hunk 12 above)
Comment (williambanfield) [reply internal/consensus/state_test.go:816]: As stated above, will do here.

**Comment by Alex-duzhichao** [comment]
@williambanfield  I have found that the implementation of tendermint is different from the algorithm description in tendermint paper too. That's confusing.  So thank you for your changes!

What I can't understand is the effect of `SkipTimeoutCommit ` parameter which doesn't been explained in tendermint paper. Why can't we enter a new height straightly but have to wait `TimeoutCommit` ? Can you give me some inspiration about `SkipTimeoutCommit `?  Thank you.
